src/audio.c

```diff
@@ -46,83 +147,127 @@ static int paudio_callback(const void *input_buffer,
 	}
+	wp = (wp + frame_count) % PA_BUFF_SIZE;
+
 	pthread_mutex_lock(&audio_mutex);
 	write_pointer = wp;
 	timestamp += frame_count;
 	pthread_mutex_unlock(&audio_mutex);
 	return 0;
 }
 
-int start_portaudio(int *nominal_sample_rate, double *real_sample_rate)
+
+void resetBuffers(){
+	pthread_mutex_lock(&audio_mutex);
+		memset(pa_buffers, 0, sizeof(pa_buffers));
+		write_pointer = 0;
+
+		timestamp = 0;
+
+	pthread_mutex_unlock(&audio_mutex);
+}
+
+int start_portaudio(int *nominal_sample_rate, double *real_sample_rate,
+					char* hostdeviceName, char*hostAPIname,
+					gboolean bShowError)
 {
-	PaStream *stream;
+
+    debug("start_portaudio %d   %s   %s \n", *nominal_sample_rate, hostAPIname, hostdeviceName);
+
 
 	if(pthread_mutex_init(&audio_mutex,NULL)) {
 		error("Failed to setup audio mutex");
 		return 1;
 	}
 
+
+
 	PaError err = Pa_Initialize();
 	if(err!=paNoError)
 		goto error;
 
-#ifdef DEBUG
-	if(testing) {
-		*nominal_sample_rate = PA_SAMPLE_RATE;
-		*real_sample_rate = PA_SAMPLE_RATE;
-		goto end;
-	}
-#endif
+	resetBuffers();
 
-	PaDeviceIndex default_input = Pa_GetDefaultInputDevice();
-	if(default_input == paNoDevice) {
-		error("No default audio input device found");
-		return 1;
+
+	PaStreamParameters inputParameters;
+	inputParameters.channelCount = 2; // Stereo
+	inputParameters.sampleFormat = paFloat32;
+	// inputParameters.suggestedLatency = ;
+	inputParameters.hostApiSpecificStreamInfo = NULL;
+	inputParameters.device = Pa_GetDefaultInputDevice();
+
+	debug("Default Input Device %d\n",inputParameters.device);
+
+
+
+
+	PaDeviceIndex selectedDeviceIndex = audio_deviceIndex(hostAPIname, hostdeviceName);
+	if(selectedDeviceIndex >= 0)
+		inputParameters.device = selectedDeviceIndex;
+
+	if(inputParameters.device == paNoDevice) {
+			error("No default audio input device found");
+			return 1;
 	}
-	long channels = Pa_GetDeviceInfo(default_input)->maxInputChannels;
+
+	const PaDeviceInfo *deviceInfo = Pa_GetDeviceInfo(inputParameters.device);
+	long long channels = deviceInfo->maxInputChannels;
 	if(channels == 0) {
-		error("Default audio device has no input channels");
+		error(" audio device has no input channels");
 		return 1;
 	}
 	if(channels > 2) channels = 2;
-	info.channels = channels;
-	info.light = false;
-	err = Pa_OpenDefaultStream(&stream,channels,0,paFloat32,PA_SAMPLE_RATE,paFramesPerBufferUnspecified,paudio_callback,&info);
+
+	if(*nominal_sample_rate == USE_DEVICE_DEFAULT_AUDIORATE){
+		*nominal_sample_rate = round(deviceInfo->defaultSampleRate);
+		if(*nominal_sample_rate > MAX_PA_SAMPLE_RATE)
+			*nominal_sample_rate = MAX_PA_SAMPLE_RATE;
+		debug("Using Device defaultSampleRate %d\n", *nominal_sample_rate);
+	}
+
+	inputParameters.channelCount = info.channels = channels;
+
+	if(inputParameters.device == Pa_GetDefaultInputDevice()){
+		debug("Pa_OpenDefaultStream dev:%d channels:%d\n",inputParameters.device, inputParameters.channelCount);
```

Comment:
What's the reason to not just use `Pa_OpenStream()` once, for any device?  Is there some difference in calling `Pa_OpenDefaultStream()`.